feat: add --index-url support for private package registries

juan-abia opened the pull request:
Add support for custom package index URLs to allow installing dependencies from private registries. The --index-url CLI option can be repeated multiple times to specify multiple indexes, which are tried in order before PyPI.

Changes:
- Add --index-url CLI argument with action='append' for multiple indexes
- Add index_urls parameter to run_mcp_server(), prepare_deno_env(), async_prepare_deno_env(), and code_sandbox()
- Pass index_urls through Deno/TypeScript layer to micropip.install()
- Update README with usage examples

Usage:
`uvx mcp-run-python --index-url https://private.repo.com/simple --deps pkg stdio`

Python API:
`async with code_sandbox( dependencies=['pkg'], index_urls=['https://private.repo.com/simple'] ) as sandbox: await sandbox.eval('import pkg')`

juan-abia committed:
feat: add --index-url support for private package registries

Add support for custom package index URLs to allow installing dependencies
from private registries. The --index-url CLI option can be repeated multiple
times to specify multiple indexes, which are tried in order before PyPI.

Changes:
- Add --index-url CLI argument with action='append' for multiple indexes
- Add index_urls parameter to run_mcp_server(), prepare_deno_env(),
  async_prepare_deno_env(), and code_sandbox()
- Pass index_urls through Deno/TypeScript layer to micropip.install()
- Update README with usage examples

Usage:
  uvx mcp-run-python --index-url https://private.repo.com/simple --deps pkg stdio

Python API:
  async with code_sandbox(
      dependencies=['pkg'],
      index_urls=['https://private.repo.com/simple']
  ) as sandbox:
      await sandbox.eval('import pkg')

DouweM commented on mcp_run_python/deno/src/runCode.ts:
```diff
@@ -122,7 +124,9 @@ export class RunCode {
-    const prepareStatus = await preparePyEnv.prepare_env(pyodide.toPy(dependencies))
+    const prepareStatus = indexUrls.length > 0
+      ? await preparePyEnv.prepare_env(pyodide.toPy(dependencies), pyodide.toPy(indexUrls))
+      : await preparePyEnv.prepare_env(pyodide.toPy(dependencies))
```
We could just always pass in the empty `indexUrls` list, right? Same as we do when `dependencies` are empty.

DouweM commented on mcp_run_python/_cli.py:
```diff
@@ -23,6 +23,13 @@ def cli_logic(args_list: Sequence[str] | None = None) -> int:
+        action='append',
+        dest='index_urls',
+        metavar='URL',
+        help='Package index URL for installing dependencies (can be repeated, tried in order before PyPI)',
```
I'd like for the deps and index URLs args to be consistent in how they take multiple values, i.e. both plural and comma separated, or both singular and repeatable. The easiest would be comma separation, although that looks a bit awkward for URLs, so we could deprecate the current `--deps` and add `--dep foo --dep bar` support as well, in line with `--index-url A --index-url B`.

juan-abia replied:
I added a new `--dep` parameter. I'm just not sure how you want to handle the deprecation.

Thanks for the feedback!

DouweM replied:
`parser.add_argument` has a `deprecated` argument :)

And please update the example in the readme to only have `--dep`, not `--deps`. Also worth seeing if there's a way to represent concisely that the args are repeatable, or if not we can mention that explicitly

juan-abia replied:
deprecated argument was added in 3.13

Should we add it with a version check? The other option is to add a warning (already added it)

juan-abia committed:
refactor: simplify indexUrls passing and add --dep flag for consistency

- Always pass indexUrls to prepare_env (even if empty), matching how dependencies are handled
- Add repeatable --dep flag alongside existing --deps for CLI consistency with --index-url
- Mark --deps as deprecated but keep it working for backwards compatibility

juan-abia committed:
fix: convert empty index_urls list to None for micropip

micropip treats empty list [] differently from None - with [] it doesn't
fall back to PyPI. Using 'index_urls or None' ensures empty lists are
converted to None so PyPI fallback works correctly.

DouweM commented on mcp_run_python/deno/src/main.ts:
```diff
@@ -51,17 +52,18 @@ Invalid arguments: ${args.join(' ')}
---return-mode <xml/json>  Return mode for output data (default: xml)`,
+--port <port>               Port to run the HTTP server on (default: 3001)
+--deps <deps>               Comma separated list of dependencies to install
+--index-urls <urls>         Comma separated list of package index URLs (tried in order before PyPI)
```
Slightly surprising they're not repeatable singular args here. I guess its not that important for these to be consistent, but see if you can make it so without too much effort

juan-abia replied:
This [commit](https://github.com/pydantic/mcp-run-python/pull/35/changes/afd0d05b5ea574a9d57d9ea30c4184fd4bdf60e7) is my attempt at making it consistent!

juan-abia committed:
docs: update CLI to use --dep instead of --deps

- Update README.md examples to use --dep (repeatable) instead of --deps
- Hide --deps from CLI help (argparse.SUPPRESS)
- Add DeprecationWarning when --deps is used
- Keep --deps working for backwards compatibility

juan-abia committed:
refactor: make Deno CLI use repeatable args for consistency

- Add --dep and --index-url as repeatable args in Deno CLI (using collect)
- Update Python main.py to pass --dep= and --index-url= instead of comma-separated
- Add deprecation warnings when --deps or --index-urls are used
- Update Deno help text to show new args
- Keep backwards compatibility for old comma-separated format
- Add test for repeatable --dep flag

juan-abia committed:
test: add tests for multiple --dep flags and deprecation warning

- test_cli_dep_multiple: test multiple --dep flags (--dep numpy --dep pydantic)
- test_cli_dep_and_deps_combined: test combining --dep and --deps
- test_cli_deps_deprecation_warning: verify deprecation warning is emitted

DouweM commented on mcp_run_python/deno/src/main.ts:
```diff
@@ -20,27 +20,45 @@ const VERSION = '0.0.13'
+    console.warn('Warning: --deps is deprecated, use --dep instead (can be repeated)')
+  }
+  if (flags['index-urls']) {
+    console.warn('Warning: --index-urls is deprecated, use --index-url instead (can be repeated)')
```
We don't need this do we? 😄

juan-abia replied:
Oups! just blindly copied from the `--deps` pattern. Removed it.

juan-abia committed:
fix: remove unnecessary --index-urls deprecation warning

--index-urls was never a public option, so there's nothing to deprecate.
